onnxruntime/core/providers/nnapi/nnapi_builtin/builders/impl/reduction_op_builder.cc

```diff
@@ -0,0 +1,222 @@
+  const bool noop_with_empty_axes = helper.Get("noop_with_empty_axes", 0);
+
+  // Get axes for ReduceMean
+  // ONNX `ReduceMean` will reduce by default all dimensions if axes is not provided/provided as empty. However, NNAPI doesn't implement the behavior
+  // to reduce all dimensions by default when 'axes' is empty/not provided. We will convert the case by providing an input with all axes for NNAPI here.
+  // Notes from NNAPI doc:
+  // https://developer.android.com/ndk/reference/group/neural-networks#group___neural_networks_1ggaabbe492c60331b13038e39d4207940e0a047fe95a35b27f45c05432b6ca18eb6c
+  std::vector<int32_t> axes;
+
+  if (node_unit.SinceVersion() >= 18) {
+    if (inputs.size() > 1 && inputs[1].node_arg.Exists()) {
+      // ReduceMean-18 uses the second optional input as axes
+      const auto& initializers(model_builder.GetInitializerTensors());
+      const auto& axes_tensor = *initializers.at(inputs[1].node_arg.Name());
+      Initializer unpacked_tensor(axes_tensor);
+      auto raw_axes = unpacked_tensor.DataAsSpan<int64_t>();
+      axes = OnnxShapeToNnapi(raw_axes, input_shape.size());
+    } else {
+      // Cases when optional axes input is not provided
+      if (noop_with_empty_axes == 0) {
+        // NNAPI doen't implement the behavior by default to reduce all dimensions when axes is empty,
+        // we will convert the case by providing an input with all axes here.
+        axes.reserve(input_shape.size());
+        std::iota(axes.begin(), axes.end(), 0);
+      }
+    }
+  } else {
+    // For ReduceMean-13 or earlier, retrieve axes from the attribute
+    const auto& axes_int64 = helper.Get("axes", std::vector<int64_t>{});
+    axes = OnnxShapeToNnapi(gsl::span<const int64_t>(axes_int64.data(), axes_int64.size()), input_shape.size());
+
+    if (axes.empty()) {
+      // When axes attribute is not specified or provided as empty, treated as default behavior to reduce all dimensions and similarly in this case,
+      // we provide an input with all axes for NNAPI here.
+      axes.resize(input_shape.size());
+      std::iota(axes.begin(), axes.end(), 0);
+    }
+  }
+
+  // Add ReduceMean operation
+  InlinedVector<uint32_t> input_indices;
+  input_indices.push_back(operand_indices.at(inputs[0].node_arg.Name()));  // data
+
+  if (!axes.empty()) {
+    const auto axes_name = model_builder.GetUniqueName(node_unit.Name() + inputs[0].node_arg.Name() + "_axes");
+    Shape axes_dimen = {static_cast<uint32_t>(axes.size())};
+    const OperandType axes_operand_type(Type::TENSOR_INT32, axes_dimen);
+    ORT_RETURN_IF_ERROR(model_builder.AddOperandFromPersistMemoryBuffer(axes_name, axes.data(), axes_operand_type));
+
+    input_indices.push_back(operand_indices.at(axes_name));  // axes
+
+    int32_t input_size = static_cast<int32_t>(input_shape.size());
+    std::vector<int32_t> axes_to_be_reduced;
+
+    for (const auto& axis : axes) {
+      axes_to_be_reduced.push_back(axis);
+    }
```

Comment:
We can use `axes` if we're not setting up an unordered_set in `axes_to_be_reduced` as they'll both be the same, and you can possibly still use the Contains helper on line 139.

Reply:
I removed `axes_to_be_reduced` here. 

but looks like the `Contains()` helper should apply to set/map?
and `axes` is a vector here. (got compile error calling Contains helper for axes.)

https://github.com/microsoft/onnxruntime/blob/cf19c3697d126982acf45c040729269e53b5d9f4/include/onnxruntime/core/common/common.h#L277-L284

Reply:
We could add a partial specialization of `Contains` to handle vectors, although that might hide accidental expensive searches of std::vector. 

I'm fine with using std::find here if we think that is something to avoid.

```c++
// partial specialization for std::vector
template <typename Key, typename LookupKey>
inline bool Contains(const std::vector<Key>& container, LookupKey&& key) {
  return std::find(container.begin(), container.end(), key) != container.end();
}
```

Reply:
one day (with C++20), I hope we can replace `Contains` with `container.contains()`. hopefully it can be a search and replace of this, so I'd prefer only having overloads that work with the std containers that have a contains member function.